File: tools/bench/answers/go_haiku/s2/14_rect_lib.go
package main

import (
	"fmt"
)

type Rectangle struct {
	width  float64
	height float64
}

func (r Rectangle) Area() float64 {
	return r.width * r.height
}

func (r Rectangle) Perimeter() float64 {
	return 2 * (r.width + r.height)
}

func (r Rectangle) Scale(factor float64) Rectangle {
	return Rectangle{r.width * factor, r.height * factor}
}

func main() {
	rect := Rectangle{3.0, 4.0}
	fmt.Println(rect.Area())
	fmt.Println(rect.Perimeter())
	scaled := rect.Scale(2.0)
	fmt.Println(scaled.Area())
}

File: tools/bench/answers/rust_haiku/s2/14_rect_lib.rs
struct Rectangle {
    width: f64,
    height: f64,
}

fn area(rect: &Rectangle) -> f64 {
    rect.width * rect.height
}

fn perimeter(rect: &Rectangle) -> f64 {
    2.0 * (rect.width + rect.height)
}

fn scale(rect: &Rectangle, factor: f64) -> Rectangle {
    Rectangle {
        width: rect.width * factor,
        height: rect.height * factor,
    }
}

fn main() {
    let rect = Rectangle {
        width: 3.0,
        height: 4.0,
    };

    println!("{}", area(&rect));
    println!("{}", perimeter(&rect));

    let scaled = scale(&rect, 2.0);
    println!("{}", area(&scaled));
}
